packages/block-editor/src/components/block-list/index.js

```diff
@@ -121,73 +136,72 @@ class BlockList extends Component {
 			return;
 		}
 
-		const boundaries = this.nodes[ clientId ].getBoundingClientRect();
-
-		// Create a clientId to Y coördinate map.
-		const clientIdToCoordMap = mapValues( this.nodes, ( node ) =>
-			node.getBoundingClientRect().top - boundaries.top );
-
-		// Cache a Y coördinate to clientId map for use in `onPointerMove`.
-		this.coordMap = invert( clientIdToCoordMap );
-		// Cache an array of the Y coördinates for use in `onPointerMove`.
-		// Sort the coördinates, as `this.nodes` will not necessarily reflect
-		// the current block sequence.
-		this.coordMapKeys = sortBy( Object.values( clientIdToCoordMap ) );
-		this.selectionAtStart = clientId;
+		this.startClientId = clientId;
+		this.props.onStartMultiSelect();
 
-		window.addEventListener( 'mousemove', this.onPointerMove );
-		// Capture scroll on all elements.
-		window.addEventListener( 'scroll', this.onScroll, true );
+		// `onSelectionStart` is called after `mousedown` and `mouseleave`
+		// (from a block). The selection ends when `mouseup` happens anywhere
+		// in the window.
 		window.addEventListener( 'mouseup', this.onSelectionEnd );
+
+		// Removing the contenteditable attributes within the block editor is
+		// essential for selection to work across editable areas. The edible
+		// hosts are removed, allowing selection to be extended outside the
+		// DOM element. `onStartMultiSelect` sets a flag in the store so the
+		// rich text components are updated, but the rerender may happen very
+		// slowly, especially in Safari for the blocks that are asynchonously
+		// rendered. To ensure the browser instantly removes the selection
+		// boundaries, we remove the contenteditable attributes manually.
+		Array.from(
+			this.ref.current.querySelectorAll( '.rich-text' )
+		).forEach( ( node ) => {
+			node.removeAttribute( 'contenteditable' );
+		} );
 	}
 
 	/**
-	 * Handles multi-selection changes in response to pointer move.
-	 *
-	 * @param {string} clientId Client ID of block under cursor in multi-select
-	 *                          drag.
+	 * Handles a mouseup event to end the current mouse multi-selection.
 	 */
-	onSelectionChange( clientId ) {
-		const { onMultiSelect, selectionStart, selectionEnd } = this.props;
-		const { selectionAtStart } = this;
-		const isAtStart = selectionAtStart === clientId;
+	onSelectionEnd() {
+		// Equivalent to attaching the listener once.
+		window.removeEventListener( 'mouseup', this.onSelectionEnd );
 
-		if ( ! selectionAtStart || ! this.props.isSelectionEnabled ) {
+		if ( ! this.props.isMultiSelecting ) {
 			return;
 		}
 
-		// If multi-selecting and cursor extent returns to the start of
-		// selection, cancel multi-select.
-		if ( isAtStart && selectionStart ) {
-			onMultiSelect( null, null );
-		}
-
-		// Expand multi-selection to block under cursor.
-		if ( ! isAtStart && selectionEnd !== clientId ) {
-			onMultiSelect( selectionAtStart, clientId );
-		}
+		this.rafId = window.requestAnimationFrame( this.setSelection );
```

Comment:
Any reason we don't do this sync?

Reply:
Yes. :) The browser selection won't have updated yet. I should leave a note...